qualtran/bloqs/mod_arithmetic/mod_multiplication.py

```diff
@@ -265,3 +273,463 @@ def _modmul_symb() -> CModMulK:
 
 
 _C_MOD_MUL_K_DOC = BloqDocSpec(bloq_cls=CModMulK, examples=(_modmul_symb, _modmul))
+
+
+@frozen
+class SingleWindowModMul(Bloq):
+    r"""Performs modular multiplication on a single windowed.
+
+    Applies
+    $$
+        \ket{x}\key{y}\key{t}\ket{0} \rightarrow \ket{x}\key{y}\ket{t+xy \mod p} \ket{xy \mod 2^w}
+    $$
+
+    Where:
+
+    - $w$ is the window size.
+    - $p$ is the modulus.
+
+    Args:
+        bitsize: size of the numbers.
+        window_size: size of the window.
+        mod: The integer modulus.
+
+    Registers:
+        x: The first integer (`window_size` bits).
+        y: The second integer (`bitsize` bits)
+        target: product accumulator.
+        qrom_index: contains the value $xy \mod 2^w$ (starts at 0).
+    """
+
+    window_size: 'SymbolicInt'
+    bitsize: 'SymbolicInt'
+    mod: 'SymbolicInt'
+
+    def __attrs_post_init__(self):
+        if not is_symbolic(self.bitsize, self.window_size):
+            assert self.bitsize % self.window_size == 0
+
+    @property
+    def signature(self) -> 'Signature':
+        return Signature(
+            [
+                Register('x', QBit(), shape=(self.window_size,)),
+                Register('y', QMontgomeryUInt(self.bitsize)),
+                Register('target', QBit(), shape=(self.window_size + self.bitsize,)),
+                Register('qrom_index', QMontgomeryUInt(self.window_size)),
+            ]
+        )
+
+    @cached_property
+    def qrom(self) -> QROAMClean:
+        if is_symbolic(self.bitsize) or is_symbolic(self.window_size) or is_symbolic(self.mod):
+            log_block_sizes = None
+            if is_symbolic(self.bitsize) and not is_symbolic(self.window_size):
+                # We assume that bitsize is much larger than window_size
+                log_block_sizes = (0,)
+            return QROAMClean(
+                [Shaped((2**self.window_size,))],
+                selection_bitsizes=(self.window_size,),
+                target_bitsizes=(self.window_size + self.bitsize,),
+                log_block_sizes=log_block_sizes,
+            )
+        inv_mod = pow(self.mod, 2 ** (self.window_size - 1) - 1, 2**self.window_size)
+        N = 2**self.window_size
+        data = (-np.arange(N) * inv_mod) % N
+        data *= self.mod
+        return QROAMClean(
+            [data],
+            selection_bitsizes=(self.window_size,),
+            target_bitsizes=(self.window_size + self.bitsize,),
+        )
+
+    def on_classical_vals(self, x: Sequence[int], y: int, target: Sequence[int], qrom_index: int):
```

Comment:
I think you _either_ need to 1) include a classical action test for this bloq or 2) remove this method so the classical action for the full DirtyModMul bloq compares the implementation against fully-tested callees.

Do you expect folks to use this `SingleWindowModMul` in other locations? If yes; then do (1). If no, then consider making this bloq private (with a preceeding underscore) and doing (2)

Reply:
I added tests ... SingleWindowModMul can be used for modular multiplication even when the number is not in montgomery form